File: best-time-to-buy-and-sell-stock-iii.cpp
//Time - O(n)
//Space - O(1)
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int buy1 = prices[0], buy2 = prices[0], profit1 = INT_MIN, profit2 = INT_MIN;
        
        for(auto p: prices){
            buy1 = min(buy1, p);
            profit1 = max(profit1, p-buy1);
            buy2 = min(buy2, p-profit1);
            profit2 = max(profit2, p-buy2);
        }
        
        return profit2;
    }
};

File: Best_Time_to_Buy_and_Sell_Stock_III.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        if not prices :
            return 0
        
        buy1, profit1, buy2, profit2 = float('inf'),0,float('inf'),0
        
        for price in prices:
            
            buy1 = min(buy1, price)
            profit1 = max(profit1, price - buy1)
            
            buy2 = min(buy2, price - profit1)
            profit2 = max (profit2, price - buy2)
            
        return profit2
        
        
